dashboard/modules/job/job_head.py

```diff
@@ -46,27 +50,123 @@
 routes = optional_utils.ClassMethodRouteTable
 
 
+class JobAgentSubmissionClient:
+    """A local client for submitting and interacting with jobs on a specific node
+    in the remote cluster.
+    Submits requests over HTTP to the job agent on the specific node using the REST API.
+    """
+
+    def __init__(
+        self,
+        dashboard_agent_address: str,
+    ):
+        self._address = dashboard_agent_address
+
+    def _do_request(
+        self,
+        method: str,
+        endpoint: str,
+        *,
+        data: Optional[bytes] = None,
+        json_data: Optional[dict] = None,
+        **kwargs,
+    ) -> requests.Response:
+        """Perform the actual HTTP request"""
+        url = self._address + endpoint
+        logger.debug(f"Sending request to {url} with json data: {json_data or {}}.")
+        return requests.request(
+            method,
+            url,
+            data=data,
+            json=json_data,
+            **kwargs,
+        )
+
+    def _raise_error(self, r: requests.Response):
+        raise RuntimeError(
+            f"Request failed with status code {r.status_code}: {r.text}."
+        )
+
+    def submit_job_internal(self, req: JobSubmitRequest) -> JobSubmitResponse:
+
+        logger.debug(f"Submitting job with submission_id={req.submission_id}.")
+
+        r = self._do_request(
+            "POST",
+            "/api/job_agent/jobs/",
+            json_data=dataclasses.asdict(req),
+        )
+
+        if r.status_code == 200:
+            return JobSubmitResponse(**r.json())
+        else:
+            self._raise_error(r)
+
+
 class JobHead(dashboard_utils.DashboardHeadModule):
     def __init__(self, dashboard_head):
         super().__init__(dashboard_head)
         self._dashboard_head = dashboard_head
         self._job_manager = None
         self._gcs_job_info_stub = None
 
-    async def _parse_and_validate_request(
-        self, req: Request, request_type: dataclass
-    ) -> Any:
-        """Parse request and cast to request type. If parsing failed, return a
-        Response object with status 400 and stacktrace instead.
+        self._disable_ray_init = dashboard_consts.RAY_RAYLETLESS_HEAD_NODE
+        # this is a queue of JobAgentSubmissionClient
+        self._agents = None
+
+    async def choose_agent(self) -> Optional[JobAgentSubmissionClient]:
```

Comment:
I feel choose agent should be a common function that can be used by other modules too. Could you abstract this part out and move it to other places?